tests/functional/adapter/snapshot_tests/test_snapshot.py

```diff
@@ -0,0 +1,205 @@
+where id between 1 and 20
+"""
+
+
+class SnapshotBase:
+
+    @pytest.fixture(scope="class")
+    def seeds(self):
+        """
+        This seed file contains all records needed for tests, including records which will be inserted after the
+        initial snapshot. This makes it so that Redshift creates the correct size varchar columns. This table
+        will only need to be loaded once at the class level. It will never be altered, hence requires no further
+        setup or teardown.
+        """
+        return {"seed.csv": seeds.SEED_CSV}
+
+    @pytest.fixture(scope="class")
+    def models(self):
+        """
+        This will be the working base table. It will be altered by each test, hence will require setup and
+        teardown at the test case level.
+        """
+        return {"fact.sql": MODEL_FACT_SQL}
+
+    @pytest.fixture(scope="class", autouse=True)
+    def _setup_class(self, project):
+        """
+        Load `seed` once for the whole class
+        """
+        run_dbt(["seed"])
+
+    @pytest.fixture(scope="function", autouse=True)
+    def _setup_method(self, project):
+        """
+        Initialize `fact` and `snapshot` for every test case.
+        Only load the first 20 `seed` records into `fact`; withhold 10 records as "new" (e.g. to test inserts).
+
+        Make the project a class variable to simplify function calls and make the code more readable.
+        For some reason this doesn't work in the class-scoped fixture, but does in the function-scoped fixture.
+        """
+        self.project = project
+        self.create_fact_from_seed("id between 1 and 20")
+        run_dbt(["snapshot"])
+        yield
+        self.delete_snapshot_records()
+        self.delete_fact_records()
+
+    def update_fact_records(self, updates: Dict[str, str], where: str = None):
+        common.update_records(self.project, "fact", updates, where)
+
+    def insert_fact_records(self, where: str = None):
+        common.insert_records(self.project, "fact", "seed", "*", where)
+
+    def delete_fact_records(self, where: str = None):
+        common.delete_records(self.project, "fact", where)
+
+    def add_fact_column(self, column: str = None, definition: str = None):
+        common.add_column(self.project, "fact", column, definition)
+
+    def create_fact_from_seed(self, where: str = None):
+        common.clone_table(self.project, "fact", "seed", "*", where)
+
+    def get_snapshot_records(self, select: str = None, where: str = None) -> List[tuple]:
+        return common.get_records(self.project, "snapshot", select, where)
+
+    def delete_snapshot_records(self):
+        common.delete_records(self.project, "snapshot")
+
+    def _assert_results(
+            self,
+            ids_with_current_snapshot_records: Iterable,
+            ids_with_closed_out_snapshot_records: Iterable
+    ):
+        """
+        All test cases are checked by considering whether a source record's id is end dated in `snapshot`. Each id
```

Comment:
nit: not sure what `end dated` means here

Reply:
Sorry, vernacular from a previous company. End dated in this context means the record's valid date is populated with the run time, thereby "closing out" the record or providing and "end date". I can update the comment with dbt-specific terminology.